txpipe/twopoint_cluster.py

```diff
@@ -0,0 +1,169 @@
+from .data_types import HDFFile, SACCFile
+
+
+class TXTwoPointCluster(PipelineStage):
+    """
+    TXPipe task for measuring two-point correlation functions
+    of galaxy clusters in an indefinite number of richness bins using TreeCorr.
+    """
+
+    name = 'TXTwoPointCluster'
+
+    inputs = [
+        ("cluster_data_catalog", HDFFile),
+        ("cluster_random_catalog", HDFFile)
+    ]
+
+    outputs = [
+        ("cluster_twopoint_real", SACCFile)
+    ]
+
+    config_options = {
+        'redshift_bin_edges': [0.4,0.8,1.2],   # Edges of redshift bins
+        'richness_bin_edges': [20, 30, 200],   # Edges of richness bins
+        'nbins': 20,                           # Number of angular bins
+        'min_sep': 0.1,                        # Minimum separation [arcmins]
+        'max_sep': 250.0,                      # Maximum separation [arcmins]
+        'units': 'arcmin',                     # Units for the separations
+        'binning_scale': 'Log'                 # 'Log', 'Linear', 'TwoD'
+    }
+
+    def load_data(self, file_path):
+        """
+        Load data from catalog (hdf5 file).
+        Read ra and dec in each richness and redshift bin.
+        """
+        # dictionaries for all the tomographic bins
+        ra  = {}
+        dec = {}
+
+        # read data and extract coordinates in each bin
+        data = h5py.File(file_path, 'r+')
+        for zbin_richbin in data['cluster_bin'].keys():
+            ra[zbin_richbin]  = data['cluster_bin'][zbin_richbin]['ra'][:]
+            dec[zbin_richbin] = data['cluster_bin'][zbin_richbin]['dec'][:]
+
+        return ra, dec 
+
+
+    def create_catalog(self, ra, dec):
+        """
+        Create a TreeCorr catalog for the given data.
+        """
+        catalog = treecorr.Catalog(ra=ra, dec=dec,
+                                   ra_units='deg', dec_units='deg')
+        return catalog
+
+
+    def measure_correlation(self, data_cat1, data_cat2,
+                                  rand_cat1, rand_cat2, bin_config):
+        """
+        Measure the cross-correlation function between two richness bins.
+        """
+        dd = treecorr.NNCorrelation(**bin_config)
+        rr = treecorr.NNCorrelation(**bin_config)
+        dr = treecorr.NNCorrelation(**bin_config)
+        rd = treecorr.NNCorrelation(**bin_config)
+
+        dd.process(data_cat1, data_cat2)
+        rr.process(rand_cat1, rand_cat2)
+        dr.process(data_cat1, rand_cat2)
+        rd.process(data_cat2, rand_cat1)
+
+        xi, varxi = dd.calculateXi(rr=rr, dr=dr, rd=rd)
+
+        return xi, varxi, dd.rnom
+
+    def save_to_sacc(self, results, output_file):
+        """
+        Save the correlation function results to a SACC file.
+        """
+        s = sacc.Sacc()
+
+        for result in results:
+            meanr, xi, varxi, tracer1, tracer2 = result
+
+            # Add tracers (placeholders z/Nz values)
+            s.add_tracer('NZ', tracer1, np.array([1]), np.array([1]))
```

Comment:
Is this just a placeholder? You can also use a "misc" tracer for this. I think each tracer here is a richness/redshift bin - is that right?  We might want to find a way to represent that.